gitbook-docs/access_requests.md

```diff
@@ -0,0 +1,97 @@
+# Access Requests
+
+When a device needs to gain access to a secured Signal K server, it can use __Access Requests__ to request and be
+granted access to the server.
+
+See [Request/Response](request_response.md) for more information on request/response semantics.
+
+A device could be a display or for example an engine sensor or a temperature sensor.
+
+## Definitions
+
+* `clientId` is a string that uniquely identifies a device. This could be a UUID, a combination of model number and
+  serial number, or any other string. The client is responsible for issuing the value. The client should use the same
+  value for all its requests.
+* `requestId` is a string generated by the server in response to an access request and used by the client to correlate
+  the request with the results. The client should not rely on the format or contents of this string.
+
+## Device Requests
+
+The device will send a REST request to the server:
+
+```sh
+$ curl -k \
+    --header "Content-Type: application/json" \
+    --request POST \
+    --data '{"clientId":"1234-45653-343453","description":"My Awesome Humidity Sensor"}' \
+    https://localhost:3443/signalk/v1/access/requests
+```
+
+If the server does not support access requests it must return HTTP status code 501 Not Implemented.
+
+For a successfully received access request the server must return an HTTP status code 202 and a JSON response with an
+href to check the status and get the response.
+
+```json
+{
+  "href": "/signalk/v1/access/requests/358b5f32-76bf-4b33-8b23-10a330827185"
+}
+```
+
+The server should then provide a process for an administrator to review and approve or deny the request.
+
+In the meantime, a device should poll the server using the `requestId` in the response above to see if it has been
+granted access and get the token.
```

Comment:
I was going to leave that up the server implementation. Do you think we need to specify it here?

Reply:
Why should the polling rate be specified? What happens if the client does not observe it?

Also access request expiration is really up to the implementation/policy imho.

Reply:
We may not be able to _prevent_ a client from spamming the server with a request every millisecond, but we can say that they don't conform to the spec if they do that. It's important to specify this kind of behavior for 3rd party vendors to follow so that later on we don't have to have the "well you didn't say I shouldn't do that" conversation when someone builds a product that effectively DoS's a server.

We don't need access request expiration _per se_. However, we do need to explicitly state what a client should do if it never gets a resolution (i.e. the request remains pending). It could be as simple as saying that if the client requests the status of an existing request and gets a 404, it should send another. But leaving it up to the servers guarantees compatibility problems in the future.

Reply:
@timmathews I updated the request/response document to include the minimum poll rate.

Reply:
> 
> We don't need access request expiration _per se_. However, we do need to explicitly state what a client should do if it never gets a resolution (i.e. the request remains pending). It could be as simple as saying that if the client requests the status of an existing request and gets a 404, it should send another. But leaving it up to the servers guarantees compatibility problems in the future.

I'm having trouble coming up with a good way to define what a client should do if never gets a resolution.

Seems like this is pretty client specific. For example, a user clicks a button to change the heading of the autopilot. In that case, the client should probably just notify the user that it doesn't know if the request worked. I don't think it should send it again. In the case of say turning on an anchor light, it would be safe to try again.

In my implementation in WilhelmSK, I just always notify the user that I do not know if the request worked. 

And maybe that's what we should state here? I don't know if the server or client will ever know if it is safe to retry a PUT request.

Reply:
In the general case PUTs should be idempotent. We can state that in the request/response page. In this specific case, a client with minimal (or no) user interface needs to have explicit rules for retrying/giving up on access requests. The case I'm concerned about is vendors building a device without explicit instructions, coming up with their own way of doing things and causing compatibility problems.

If we give them detailed rules to follow, hopefully we can avoid that.

Reply:
I can't quite grasp what the compatibility problems would be. Elaborate, please.

Neither can I think of universal rules to follow. A device with no rtc can not rely on minimum delays. In a scenario where the device has submitted one and the server has misplaced it (reinstalling the server software for example and wiping persistent state) there should be some simple way to trigger rerequesting access. Then there are timing issues (both server and device on the same power, either device or

Reply:
Compatibility problems arise when both sides are unsure of what to do. I cannot predict what they might be, but I can suggest how to minimize the potential for them.

>  A device with no rtc can not rely on minimum delays.

Client devices do not need an RTC to count milliseconds, they absolutely can be told to not GET their status more often than every 500ms. This is specified now in the documentation on PUT requests, so I think we're OK here.

>  In a scenario where the device has submitted one and the server has misplaced it (reinstalling the server software for example and wiping persistent state) there should be some simple way to trigger rerequesting access.

As long as the client can see that the request is expired or denied, they can re-request. This just needs to be stated explicitly.

Here are a couple of scenarios where I think we need to define what the client and server actions should be.

1. A device successfully pairs with a server and the server is replaced[1]. When it connects to the new server, does it a) fail silently or b) upon receipt of an `unauthorized` message respond with a new access request? It seems obvious to me that it should be **b**, but I doubt that it will be obvious to every implementor unless spelled out.

2. If a device is not permitted access, when is it OK to ask again? Does it keep asking forever, or just once at boot-up?

3. When access is revoked, what does the client device do?

[1] Either the user sets up a new server or they sell the client device. I'm sure there are more cases when this could occur, but "replaced" is fine in the general case.

Reply:
@timmathews I think I've address all your use cases.